app-creator/src/widgets/ui-map/ui-map.ts

```diff
@@ -37,106 +46,161 @@ function loadGoogleMaps(apiKey: string) {
-   */
-  @property({ type: Object }) styles = {};
-
   /**
    * Map api key.
    */
-  @property({ type: String }) apiKey = '';
+  private apiKey = '';
 
   /**
    * Sets zoom property on map.
    */
-  @property({ type: Number }) zoom = 4;
+  private zoom = 4;
 
   /**
    * Center position of map.
    */
-  @property({ type: Object }) center = {
-    lat: 39.930546488601294,
-    lng: -100.825,
+  private center = {
+    lat: 37.419857,
+    lng: -122.078827,
   };
 
+  /**
+   * If set to true, adds street view button.
+   */
+  private streetViewControl = false;
+
+  /**
+   * Enables zoom control on map.
+   */
+  private zoomControl = false;
+
+  /**
+   * If set to true, adds full screen button.
+   */
+  private fullscreenControl = false;
+
+  /**
+   * If set to true, adds map type control.
+   */
+  private mapTypeControl = false;
+
+  /**
+   * Sets pre-existing map styling.
+   * Options: aubergine, dark, night, retro, silver, standard.
+   */
+  private mapStyles = standard;
+
+  /**
+   * Sets map styling to custom JSON.
+   * Custom JSON can be generated from https://mapstyle.withgoogle.com/
+   */
+  private customMapStyles = '';
+
   /**
    * Properties of map. Full list can be found here.
    * https://developers.google.com/maps/documentation/javascript/reference/map#MapOptions.
    */
-  @property({ type: Object }) mapOptions: google.maps.MapOptions = {};
+  private mapOptions: google.maps.MapOptions = {};
 
   /**
    * Sets map element.
    */
   private map: google.maps.Map | null = null;
 
-  /**
-   * Adds edit icon to panel.
-   */
-  @property({ type: Boolean }) editable = false;
-
   connectedCallback() {
-    this.initMap();
+    this.onmousedown = this.handleMouseDown;
+    this.onclick = (e: Event) => e.stopPropagation();
+  }
+
+  handleMouseDown(e: Event) {
+    e.stopPropagation();
+    DraggableWidget.removeEditingWidgetHighlight();
+    if (store.getState().editingElement != this) {
+      store.dispatch(setEditingWidget(this as Map));
+    }
+    store.dispatch(setSelectedTab(Tab.attributes));
   }
 
-  initMap() {
+  /**
+   * Called on initialization and on property changes. Initial call
+   * creates a new map instance but subsequent ones update the existed object.
+   */
+  updateMap() {
     loadGoogleMaps(this.apiKey).then(() => {
       this.mapOptions.zoom = this.zoom || 0;
 
       this.mapOptions.center = this.center;
 
-      this.mapOptions.zoomControl = false;
+      this.mapOptions.zoomControl = this.zoomControl;
+
+      this.mapOptions.streetViewControl = this.streetViewControl;
 
-      this.mapOptions.streetViewControl = false;
+      this.mapOptions.fullscreenControl = this.fullscreenControl;
 
-      this.mapOptions.fullscreenControl = false;
+      this.mapOptions.mapTypeControl = this.mapTypeControl;
 
-      this.mapOptions.mapTypeControl = false;
+      this.mapOptions.styles =
+        this.customMapStyles != ''
+          ? JSON.parse(this.customMapStyles)
```

Comment:
If the user inputs an invalid JSON, this would throw an error. Is data validation handled somewhere?

Reply:
Added an optional Validator function to the attribute interface.
```
export interface AttributeMetaData {
  [key: string]: {
    value: string;
    placeholder?: string;
    ...
    validator?: Function;
  };
}
```

Then in ui-map, we pass in a function like so: 
```
customMapStyles: {
      value: '',
      validator: (value: string) => {
        try {
          JSON.parse(value);
          return true;
        } catch (e) {
          return false;
        }
      },
      placeholder: 'Paste JSON here',
      type: InputType.textarea,
      tooltip: {
        text: 'Create custom map styles at the following link.',
        url: 'https://mapstyle.withgoogle.com',
      },
    },
```

Lastly, on textAreaKeyup in the attributes-tab, we check if the value is valid: 
```
  handleTextAreaKeyup(
    ...
    validator?: Function
  ) {
    const textareaInput = e.target as HTMLTextAreaElement;
    if (validator != null) {
      if (validator(textareaInput.value) || textareaInput.value === '') {
      ...
      } else {
        ...
      }
    } else {
      ...
    }
  }
```